### DISCUSSION TITLE
Prisma middleware ignored in tests

### DISCUSSION THREAD

**Discussion by notflip** [prisma/prisma#19806] [Q&A]
I'm following the setup from the docs for testing: https://www.prisma.io/docs/guides/testing/unit-testing

This is my client.ts
```ts
import { PrismaClient } from "@prisma/client";
import { DateTime } from "luxon";

const prisma = new PrismaClient();

// Middleware
prisma.$use(async (params, next) => {
    console.log('------ middleware running -------') // <- this console.log is never shown
    if (params.model == 'LonenAppl' && params.action == 'create') {
        const originalDatumAfdruk = params.args.data.datum_afdruk
        params.args.data.datum_afdruk = DateTime.fromFormat(originalDatumAfdruk, 'ddMMyy')
            .setZone('Europe/Brussels')
            .toISODate()
    }

    return next(params)
})

export default prisma;
```

I've got the correct singleton.ts, the console.log here is shown during tests
```ts
import { PrismaClient } from '@prisma/client'
import { mockDeep, mockReset, DeepMockProxy } from 'jest-mock-extended'

import prisma from './client'

jest.mock('./client', () => ({
    __esModule: true,
    default: mockDeep<PrismaClient>(),
}))

beforeEach(() => {
    console.log('------- beforeEach running ------')
    mockReset(prismaMock)
})

export const prismaMock = prisma as unknown as DeepMockProxy<PrismaClient>
```

Any idea on how I can include the middleware inside the unit tests?

**Comment by nurul3101** [comment]
Hello @notflip 👋 

Since you are using `jest.mock` to mock the `PrismaClient`, the actual implementation of the PrismaClient (including the middleware) is not being used in the tests. This is why I believe the `console.log` inside the middleware is never shown during the tests.

Also when you are unit testing, typically you would mock out external dependencies and focus on testing the logic of the unit (function, module, etc.) in isolation. If you are specifically looking to test the interaction with the database and the middleware logic, this would be more akin to integration testing. In integration testing, it's common to have a real instance of the database running (often in-memory), and the tests would verify that the queries, middleware, and other logic work as expected together..

**Reply by notflip** [reply]
That makes sense, thanks for clarifying
